File: tasks/text-redact/javascript/solution/parse.js
"use strict";

const TOP_LEVEL = ["config", "rules", "text"];
const CONFIG_KEYS = ["mask", "minLength", "policy"];
const LITERAL_KEYS = ["id", "kind", "value"];
const SPAN_KEYS = ["end", "id", "kind", "start"];
const POLICIES = ["merge", "strict"];
const RULE_ID = /^[A-Za-z0-9_.-]+$/;

function isObject(value) {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function hasKeys(value, keys) {
  return Object.keys(value).sort().join(",") === keys.join(",");
}

function isInteger(value) {
  return typeof value === "number" && Number.isInteger(value);
}

/** The text as an array of Unicode code points. */
function codePoints(text) {
  return Array.from(text);
}

function parseDocument(value) {
  if (!isObject(value) || !hasKeys(value, TOP_LEVEL)) {
    throw new Error("malformed document");
  }
  if (typeof value.text !== "string") {
    throw new Error("malformed text");
  }
  if (!Array.isArray(value.rules)) {
    throw new Error("malformed rules");
  }
  return value;
}

function parseConfig(value) {
  if (!isObject(value) || !hasKeys(value, CONFIG_KEYS)) {
    throw new Error("malformed config");
  }
  const { mask, policy, minLength } = value;
  if (typeof mask !== "string" || codePoints(mask).length !== 1) {
    throw new Error("malformed mask");
  }
  if (typeof policy !== "string" || !POLICIES.includes(policy)) {
    throw new Error("malformed policy");
  }
  if (!isInteger(minLength) || minLength < 1) {
    throw new Error("malformed minimum length");
  }
  return { mask, policy, minLength };
}

function parseRule(value, seen, length) {
  if (!isObject(value) || !("kind" in value)) {
    throw new Error("malformed rule");
  }
  const id = value.id;
  if (typeof id !== "string" || !RULE_ID.test(id)) {
    throw new Error("malformed rule id");
  }
  if (seen.has(id)) {
    throw new Error("duplicate rule id");
  }
  if (value.kind === "literal") {
    if (!hasKeys(value, LITERAL_KEYS)) {
      throw new Error("malformed literal rule");
    }
    if (typeof value.value !== "string" || value.value.length === 0) {
      throw new Error("malformed literal value");
    }
    return { id, kind: "literal", value: value.value };
  }
  if (value.kind === "span") {
    if (!hasKeys(value, SPAN_KEYS)) {
      throw new Error("malformed span rule");
    }
    const { start, end } = value;
    if (!isInteger(start) || !isInteger(end) || start < 0) {
      throw new Error("malformed span bounds");
    }
    if (start >= end || end > length) {
      throw new Error("malformed span bounds");
    }
    return { id, kind: "span", start, end };
  }
  throw new Error("unknown rule kind");
}

module.exports = { codePoints, parseConfig, parseDocument, parseRule };

File: tasks/text-redact/go/solution/parse.go
package main

import (
	"encoding/json"
	"errors"
	"regexp"
	"slices"
	"strconv"
)

var ruleIDPattern = regexp.MustCompile(`^[A-Za-z0-9_.-]+$`)

var (
	topLevel    = []string{"config", "rules", "text"}
	configKeys  = []string{"mask", "minLength", "policy"}
	literalKeys = []string{"id", "kind", "value"}
	spanKeys    = []string{"end", "id", "kind", "start"}
)

type config struct {
	Mask      string
	Policy    string
	MinLength int
}

type rule struct {
	ID    string
	Kind  string
	Value string
	Start int
	End   int
}

type document struct {
	Config any
	Text   string
	Rules  []any
}

func hasKeys(value map[string]any, keys []string) bool {
	if len(value) != len(keys) {
		return false
	}
	present := make([]string, 0, len(value))
	for key := range value {
		present = append(present, key)
	}
	slices.Sort(present)
	return slices.Equal(present, keys)
}

// asInt reports the integer value of a decoded JSON number.
func asInt(value any) (int, bool) {
	number, ok := value.(json.Number)
	if !ok {
		return 0, false
	}
	parsed, err := strconv.Atoi(number.String())
	if err != nil {
		return 0, false
	}
	return parsed, true
}

// codePoints returns the text as a slice of Unicode code points.
func codePoints(text string) []rune {
	return []rune(text)
}

func parseDocument(value any) (document, error) {
	object, ok := value.(map[string]any)
	if !ok || !hasKeys(object, topLevel) {
		return document{}, errors.New("malformed document")
	}
	text, ok := object["text"].(string)
	if !ok {
		return document{}, errors.New("malformed text")
	}
	rules, ok := object["rules"].([]any)
	if !ok {
		return document{}, errors.New("malformed rules")
	}
	return document{Config: object["config"], Text: text, Rules: rules}, nil
}

func parseConfig(value any) (config, error) {
	object, ok := value.(map[string]any)
	if !ok || !hasKeys(object, configKeys) {
		return config{}, errors.New("malformed config")
	}
	mask, ok := object["mask"].(string)
	if !ok || len(codePoints(mask)) != 1 {
		return config{}, errors.New("malformed mask")
	}
	policy, ok := object["policy"].(string)
	if !ok || (policy != "merge" && policy != "strict") {
		return config{}, errors.New("malformed policy")
	}
	minimum, ok := asInt(object["minLength"])
	if !ok || minimum < 1 {
		return config{}, errors.New("malformed minimum length")
	}
	return config{Mask: mask, Policy: policy, MinLength: minimum}, nil
}

func parseRule(value any, seen map[string]bool, length int) (rule, error) {
	object, ok := value.(map[string]any)
	if !ok {
		return rule{}, errors.New("malformed rule")
	}
	if _, present := object["kind"]; !present {
		return rule{}, errors.New("malformed rule")
	}
	identifier, ok := object["id"].(string)
	if !ok || !ruleIDPattern.MatchString(identifier) {
		return rule{}, errors.New("malformed rule id")
	}
	if seen[identifier] {
		return rule{}, errors.New("duplicate rule id")
	}
	kind, ok := object["kind"].(string)
	if !ok {
		return rule{}, errors.New("malformed rule")
	}
	if kind == "literal" {
		if !hasKeys(object, literalKeys) {
			return rule{}, errors.New("malformed literal rule")
		}
		text, ok := object["value"].(string)
		if !ok || text == "" {
			return rule{}, errors.New("malformed literal value")
		}
		return rule{ID: identifier, Kind: "literal", Value: text}, nil
	}
	if kind == "span" {
		if !hasKeys(object, spanKeys) {
			return rule{}, errors.New("malformed span rule")
		}
		start, startOK := asInt(object["start"])
		end, endOK := asInt(object["end"])
		if !startOK || !endOK {
			return rule{}, errors.New("malformed span bounds")
		}
		if start < 0 || start >= end || end > length {
			return rule{}, errors.New("malformed span bounds")
		}
		return rule{ID: identifier, Kind: "span", Start: start, End: end}, nil
	}
	return rule{}, errors.New("unknown rule kind")
}
